### PR TITLE
profile install: warn on installing package twice

### PR DESCRIPTION
**Pull request by bobvanderlinden**
# Motivation

Currently it is possible to install a package twice.

```console
$ nix profile install nixpkgs#firefox
$ nix profile install nixpkgs#firefox
$ nix profile list
...
Name: firefox
...
Name: firefox-1
```

This change explicitly checks for duplicate entries when installing packages. It'll look like:

```console
$ nix profile install nixpkgs#firefox
$ nix profile install nixpkgs#firefox
error: 'firefox' is already installed
```

To decide a duplicate, the name, `attrPath` and `originalRef` (`flake:nixpkgs`) should be the same.

`lockedRef` (`github:NixOS/nixpkgs/2a34566b67bef34c551f204063faeecc444ae9da`) is ignored. This is to make sure installing the package twice at any time in the future will still be a conflict.

# Context

Fixes https://github.com/NixOS/nix/issues/5587
As suggested in https://github.com/NixOS/nix/pull/10065#issuecomment-1961285593

# Priorities and Process

Add :+1: to [pull requests you find important](https://github.com/NixOS/nix/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc).

The Nix maintainer team uses a [GitHub project board](https://github.com/orgs/NixOS/projects/19) to [schedule and track reviews](https://github.com/NixOS/nix/tree/master/maintainers#project-board-protocol).

### CODE REVIEW

File: src/nix/profile.cc [hunk 1]
```diff
@@ -390,7 +395,25 @@ struct CmdProfileInstall : InstallablesCommand, MixDefaultProfile
...
+                auto existingElement = existingPair->second;
+                auto existingSource = existingElement.source;
+                auto elementSource = element.source;
+                if (existingSource
+                    && elementSource
+                    && existingElement.priority == element.priority
+                    && existingSource->originalRef == elementSource->originalRef
+                    && existingSource->attrPath == elementSource->attrPath
+                    ) {
+                    throw Error("'%s' is already installed", elementName);
```
Comment (edolstra) [inline src/nix/profile.cc:412]: Maybe repeated installation should be idempotent, i.e.
```suggestion
                    warn("'%s' is already installed", elementName);
                    continue;
```

This way, `nix profile install X` is defined in terms of the post-condition being that X is installed. That could be convenient for e.g. installation scripts.

Comment (bobvanderlinden) [reply src/nix/profile.cc]: Ah, yes I can see being less strict would help :+1: I've applied your suggestion. Thanks!

**Review by thufschmitt — APPROVED** [review]

**Review by edolstra — APPROVED** [review]

**Comment by maksbotan** [comment]
Hi! Would there be a way to make `nix profile` replace old instance of package?

For example, if I'm installing a tool from local repository with `default.nix`, and I would like to reinstall it when I update the repo.

**Comment by bobvanderlinden** [comment]
It may be a good idea to create an issue. I can take a look at it if there is a concrete plan how that workflow should work. I _think_ a separate flag (like `--force`) is need for something like this.

**Pull request merged**